Complete the API reference docs

FirefoxMetzger opened the pull request:
Closes: https://github.com/pygfx/pygfx/issues/427

This PR adds missing API docs and begins to complete https://github.com/pygfx/pygfx/issues/113.

almarklein commented on docs/reference.rst:
```diff
@@ -1,13 +1,63 @@
+
+.. autosummary::
+
+    pygfx.objects.WorldObject
+    pygfx.utils.color.Color
```
Can we name these just `pygfx.WorldObject` etc?

almarklein replied:
> It should be possible to create alias names or get autosummary to show a different name in the result

That would be great. That `pygfx.utils.color.Color` is rather ugly in hindsight 😄

FirefoxMetzger replied:
I looked into this over the weekend and there are two options:

1. We can hide the module in the autosummary table. This would turn `pygfx.utils.color.Color` into `Color`. Autosummary supports this out-of-the-box, and this is how I solved it for now.
2. We can add a small custom sphinx extension that extends autosummary by adding a custom prefix to the name in the table, i.e., it would turn `pygfx.utils.color.Color` into `pygfx.Color`.

In both cases the result would link to the same page that we use in the sub-package docs, so either solution works. Do you have any preference @almarklein ?

almarklein replied:
Let's use the builtin option and use `Color`.

FirefoxMetzger commented:
FYI @almarklein I began adding a mini gallery to the docstrings that I am editing. We briefly discussed this once during a meeting, but never dug deeper into it.

You can now see rendered examples of them here:
- `box_geometry`: https://pygfx--430.org.readthedocs.build/en/430/_autosummary/_autosummary/pygfx.geometries.box_geometry.html#pygfx.geometries.box_geometry
- `trimesh_geometry` https://pygfx--430.org.readthedocs.build/en/430/_autosummary/_autosummary/pygfx.geometries.trimesh_geometry.html#pygfx.geometries.trimesh_geometry

FirefoxMetzger commented:
@almarklein I just realized that I began to instinctively turn all the docstrings that I found into numpy/numpydoc style docstrings. However, I just realized that we currently use a mix of google-style and numpy style in that it mostly follows google's convention, but uses `Parameters` instead of `Args`. On the other hand, pylinalg imposes numpydoc style.

I didn't find any explicit statement on which kind of style to go for and was wondering if you have a strong opinion either way. (Since I am going over most of the codebase anyway to update docstrings we could combine this with a style change if desired or add an explicit comment on this somewhere.)

Korijn commented:
Personally I prefer numpydoc, but on the other hand I've never done anything this extensive with sphinx so feel free to opt for anything else you find more productive.

FirefoxMetzger commented:
I, too, lean towards the numpydoc style, due to the added flexibility provided by the increased number of sections when compared to google's style.

At the same time, @almarklein is writing a lot of the code for this project, so I think he should also weigh in on this.

almarklein commented:
In general I have a minor preference for the Google style because it feels lighter. But for this project I agree that numpydoc would probably be more suitable.

FirefoxMetzger commented:
@almarklein I'm done with the geometry submodule. If you like, feel free to review it.

FirefoxMetzger commented:
The helper module is done as well.

FirefoxMetzger commented:
@Korijn @almarklein what shall we do with the linalg module? It will (hopefully soon) be removed in favor of pylinalg, which makes it a bit redundant to document it extensively. I can, ofc do that if you think it has merit, but personally I would rather put a warning on the module and say "this module will disappear soon. check out pylinalg instead".

almarklein commented:
>  I would rather put a warning on the module and say "this module will disappear soon. check out pylinalg instead".

Yes, let's not waste effort on documenting the to-be-deprecated linalg.

almarklein commented on pygfx/helpers/_box.py:
```diff
@@ -4,14 +4,29 @@
 
+    # @almarklein: Should we move this to geometry._box? It clearly has a
+    # usecase as a aabb helper, but seems to have merit outside of this context,
+    # too.
```
I don't have a strong opinion on the module that this is in, but its still be a helper, so this module makes sense? 😃

FirefoxMetzger replied:
We could call it `WireframeBox` and then the name "Helper" disappears ;). I mainly think that a box wireframe can do more than just help draw a bounding box.

(also while on the topic of wireframes, shouldn't it be possible to write a shader that draws a wireframe instead of drawing faces, removing the need for explicit wireframe objects?)

almarklein replied:
> We could call it `WireframeBox` and then the name "Helper" disappears ;). I mainly think that a box wireframe can do more than just help draw a bounding box.

I'm not convinced yet. It still feels like a helper to me 😄. If you take your position to the extreme, you could argue that all objects in the helper submodule should be first class world objects. In the end it does not matter much where they live. Let's discuss in a new issue?

> (also while on the topic of wireframes, shouldn't it be possible to write a shader that draws a wireframe instead of drawing faces, removing the need for explicit wireframe objects?)

Yes, we have that! It's the `MeshMaterial.wireframe` property. But you'll see wires for all the edges of the triangle, so it serves a different purpose.

Korijn replied:
The qualitative difference between geometry and helper is that helpers are WorldObjects and geometries are routines that produce geometry which can be attached to WorldObjects.

In this specific case, if anything, you could extract the parts of code that generate the box's geometry to a function in the geometry subpackage. But the box helper will remain.

FirefoxMetzger replied:
> Yes, we have that! It's the MeshMaterial.wireframe property. But you'll see wires for all the edges of the triangle, so it serves a different purpose.

Ah right, then it isn't equivalent to drawing a wireframe box as we do here.

> WorldObjects typically used for visual debugging, while the geometry package holds routines that produce geometry which can be attached to WorldObjects

Ah, that's the piece I was missing. It will be good to have this info in the docs of the helpers module, because it makes it clear why the box helper is relevant here. I'll go and update the docstring.

almarklein commented on pygfx/helpers/_box.py:
```diff
@@ -46,27 +61,34 @@ def __init__(self, size=1.0, thickness=1, color="white"):
         )
         positions -= 0.5
+
+        # @almarklein: does this mean we currently don't support non-isotropic scale?
+        # otherwise we may want to update this to set the tranform's scale instead.
```
The geometry is always a cube, but the transform's scale can be used to stretch it. See `self.scale.set` below.

FirefoxMetzger replied:
In that case, why do we need `scale` _and_ `size`? Wouldn't it be enough to always construct a unit cube and use scale?

almarklein replied:
Good point. I suspect that it has a `size` arg because many geometry-generating functions have it. And for these it makes sense, because `.scale` also sizes the children and you may not want to do that. But helper objects are unlikely to have children. I'd be ok with removing the `size` arg. @Korijn ?

Korijn replied:
As mentioned in the other open conversation on _box.py, there is an inline geometry routine in the constructor here. That routine is parameterized so that users can generate whatever geometry they need. I guess it would make sense actually to also add subdivision parameters, since users may have UV mapping use cases for which they require more triangles on each side of the cube. But those are all concerns for a box geometry generation routine.

For the helper object, it doesn't matter too much, since indeed, you also have the standard scale/rotation/position transform options at your disposal. I ported these from three.js for the most part, and I assume the arguments are just there for one-liner visual debugging convenience. I would prefer the size arg to stay for this reason.

FirefoxMetzger replied:
@Korijn  Is it because of the name, or because of the ability to use `size` and `scale` independently? If the former, we could look into making `size` an alias of `scale` which should make our code a little less complex while maintaining the current behavior.

Korijn replied:
I'm assuming there's no specific conscious design decision behind this.

almarklein commented on pygfx/helpers/_box.py:
```diff
@@ -46,27 +61,34 @@ def __init__(self, size=1.0, thickness=1, color="white"):
 
         super().__init__(geometry, material)
 
+    # should this become a classmethod? (alternative constructor)
```
I don't think so. With this method you can have a single helper object, and use it to "highlight" objects in the scene, e.g. as the user clicks them.

FirefoxMetzger replied:
Ah, I see.

almarklein commented on pygfx/helpers/_gizmo.py:
```diff
@@ -297,10 +326,19 @@ def _highlight(self, object=None):
+
         """
+
+        # @almarklein: Assuming I interpreted above docstring correctly,
+        # wouldn't it be easier to make the gizmo a child of the controlled
+        # object and let the transform/linalg engine take care of the moving?
+        # (as its will move together with its parent, who's position the gizmo
+        # updates.)
```
We tried that at the time, but decided against it. I can't recall the reasons though 🤔

FirefoxMetzger replied:
So we keep things as-is, or should we make an issue on attempting to change this (and shed some code in the process)?

almarklein replied:
Let's keep it as-is. Tried to remember a bit harder. The thing is that if you make it a child of the object, you must undo the scaling of the world object, and depending on the mode, also the rest of the transform. This would make the calculations unnecessarily complex, defeating the purpose of using the parent-child relationship for selecting the object to control.

almarklein commented on setup.py:
```diff
@@ -46,6 +46,12 @@
         "sphinx-gallery",
         "imageio",
         "matplotlib",
+        # duplicate example dependencies to avoid pyside6
```
Maybe PySide6 should not be in the example deps. A few examples are qt-specific, but it can also be e.g. pyqt5. @Korijn ?

Korijn replied:
Can you explain why we want avoid pyside6?

FirefoxMetzger replied:
@Korijn because we don't need it when building the docs (keeps build times low), and because it started to break the build on RTD.

Korijn replied:
OK.

almarklein commented on pygfx/geometries/_box.py:
```diff
@@ -14,24 +14,43 @@ def box_geometry(
+    number of segments along each axis. This will result in all faces parallel to
+    the given axis to be evenly devided into the requested number of segments.
+
+    .. @almarklein: What does it mean for corners to be square?
+
+    The normals for each side all point in the same direction; the corners are
+    actually square. Texture coordinates for each side run from 0 to 1.
+
```
Meh, let's drop this comment. The thought was that if you'd share the corner vertices, they'd appear oddly rounded. But this comments feels out of place now.
```suggestion
```

almarklein commented on pygfx/helpers/_gizmo.py:
```diff
@@ -494,21 +532,25 @@ def _update_visibility(self):
         self._viewport = viewport
         self._camera = camera
 
+        # @almarklein: out of curiosity, what happens if we have more than one
+        # gizmo? Will events alias/collide?
```
Unless I'm mistaken, the gizmo only gets events when its elements are clicked, so it *should* work.

FirefoxMetzger replied:
So each gizmo has it's own event queue? It makes sense that only the gizmo that is clicked emits an event, but I don't know how it works on the receiving end if multiple gizmos are listening.

almarklein replied:
Events are only dispatched to world objects if they are clicked on (or hovered over, etc.).

In more detail: the canvas generates the event. The renderer picks it up and checks if it applies to a certain world object, e.g. because the object is under the mouse, or when a world object is tracked (i.e. dragging). And the event propagates to parent world objects, similar to how it does in the browser dom.

Korijn replied:
The event system is modelled after the browser dom and is pretty complex, but also very flexible and powerful :) it's one of pygfx' killer features, I would say. I am not familiar with anything this easy to use in other graphics libraries for python.

FirefoxMetzger replied:
> Events are only dispatched to world objects if they are clicked on (or hovered over, etc.).

This means the object's geometry right? Like, we cast a ray through the cursor's pixel position and the first geometry it hits is the one that generates events? In other words, objects without geometries won't receive events?

> And the event propagates to parent world objects, similar to how it does in the browser dom.

Oh, that's a really cool way to isolate events. 

Since it uses the parent/child relationship it does make me wonder though what this means for the new objects API and our plans of moving the parent/child relationship into `AffineTransform`. Our argument for this was that we only use the scene graph for transforms and that chaining should be testable in pylinalg. Since we are using it for events, too, does this change anything?

Korijn replied:
> This means the object's geometry right? Like, we cast a ray through the cursor's pixel position and the first geometry it hits is the one that generates events? In other words, objects without geometries won't receive events?

Actually, no, all shaders render both a color pixel to a buffer, and picking info to a secondary buffer. We query that secondary buffer for picking information. It includes different things depending on the material used to render. For example, here is the picking implementation for Mesh:

* https://github.com/pygfx/pygfx/blob/4aafa96c1f7ac21d6d2b1a177a6a44dcb5779daa/pygfx/materials/_mesh.py#L39
* https://github.com/pygfx/pygfx/blob/4aafa96c1f7ac21d6d2b1a177a6a44dcb5779daa/pygfx/renderers/wgpu/meshshader.py#L396-L405

> Since it uses the parent/child relationship it does make me wonder though what this means for the new objects API and our plans of moving the parent/child relationship into `AffineTransform`. Our argument for this was that we only use the scene graph for transforms and that chaining should be testable in pylinalg. Since we are using it for events, too, does this change anything?

We'll need to be creative and find a way to satisfy all our design goals. Must haves for me are still the same:

* chaining should be testable in pylinalg
* parent/child relationships should be user friendly, traversable and queryable in pygfx

almarklein commented on pygfx/helpers/_lights.py:
```diff
@@ -149,10 +182,28 @@ def _update(self):
+
     """
 
+    # @almarklein: What is the expected shape of this indicator? Also, could we
+    # combine its code path with that for the spot light, but make the Mesh a
+    # cone that points into the opposite direction of the light source (ie. in
+    # such a way that the bottom of the cone has the light coming out, like a
+    # lamp)
```
They look like this:
<img width="566" alt="image" src="https://user-images.githubusercontent.com/3015475/212285624-8275966a-d9d6-4af8-a216-846e034498c3.png">

The idea of the helper to see quite precisely where it's pointed at, which would not be as clear with the "lamp". It could be worth improving the helper, but it'd require more discussion, so it'd be more suited in a separate issue/PR.

FirefoxMetzger replied:
That makes sense. In this case, lets take it up again once there is a specific need/use-case for an improvement.

almarklein commented on pygfx/geometries/_cylinder.py:
```diff
@@ -143,7 +143,53 @@ def cylinder_geometry(
-    """Create geometry representing a cylinder."""
+    """Generate a cylinder or a cylinder segment.
+
+    .. @almarklein @korjin: isn't this inconsistend
+    .. with our convnetion that the z-axis is forward?
+
+    This function generates a cylinder or a cylinder segment. The cylinder's
+    axis runs along the local z-axis, and its midpoint is located at the local
+    origin. The cylinder's faces (top and bottom cap) are approximated by
+    regular N-sided polygons, with corners on a circle of the given radius.
```
>  The cylinder's axis runs along the local z-axis

I don't think this is the case. I ran the `speres.py` example, and modify the geometry like this:
```py
geometry.positions.data[:,1] *= 0.5
```
Causing the spheres to be flat in the direction of their axis.
<img width="587" alt="image" src="https://user-images.githubusercontent.com/3015475/212287443-d6345118-6834-4e33-98b4-d8b618345551.png">

FirefoxMetzger replied:
I got this from the source of the cylinder, which constructs the mantle positions using the following segment:

https://github.com/pygfx/pygfx/blob/eb085bd5010445ea7ddc807da0d4e7a012a6798c/pygfx/geometries/_cylinder.py#L36-L39

I didn't do any testing on this though, so it could be that there is a rotation somewhere that matches it with our convention.

almarklein replied:
Oh, but the cylinders are indeed not facing up:
<img width="481" alt="image" src="https://user-images.githubusercontent.com/3015475/212412973-3f76228c-fd9a-45b8-b804-2cce7996993a.png">

I think this is something that we should change.

FirefoxMetzger replied:
Let's discuss this in https://github.com/pygfx/pygfx/issues/436

almarklein reviewed (commented):
Nice work so far!

FirefoxMetzger commented on docs/conf.py:
```diff
@@ -48,7 +48,13 @@ def _check_objects_are_documented():
+# Also: this breaks when we autogenerate the docs, because there will be no
+# explicit `.. autoclass::` line anymore. I could work out how to adapt this
+# function though, if you think it is worth keeping.
+# _check_objects_are_documented()
```
You still have to manually "mention" a new function in the autosummary list of the respective module. It's harder to miss, because you will see the module's docstring when you add the function to the module's `__init__.py`, but one could still miss it. (We could automate this, too, but it significantly reduces our flexibility in writing nice docs.)

That said, I'm not aware of any other repo that has such a check, so keeping/discarding depends on how "bad" we think it is if we miss a function in the docs. (We can always add it later and the docs update instantly.)

almarklein replied:
Anyway, let's drop the check here.

FirefoxMetzger commented on pygfx/helpers/_axes.py:
```diff
@@ -13,11 +13,23 @@
 
 class AxesHelper(Line):
-    """An object indicating the axes directions.
+    """Generate an axis indicator.
```
This first sentence is what shows up in the summary table of the module, so I like to write them in a way that they are coherent when you see them all aggregated there. (ref: https://pygfx--430.org.readthedocs.build/en/430/_autosummary/pygfx.geometries.html)

As I understand it, [Numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#documenting-classes) recommends documenting the constructor's arguments in the classes docstring (under `Parameters`). Since we do that, I think it makes sense to match the rest of the style as if we are documenting the constructor.

almarklein replied:
> This first sentence is what shows up in the summary table of the module, so I like to write them in a way that they are coherent when you see them all aggregated there.

Fair enough for those use-cases.

> Since we do that, I think it makes sense to match the rest of the style as if we are documenting the constructor.

From my experience it's more common to describe the object. The fact that you also describe the parameters to instantiate such an object does not conflict IMO. This is an example from the same numpydoc page:

<img width="445" alt="image" src="https://user-images.githubusercontent.com/3015475/212407998-e76eb112-0cf8-47ba-b5ca-8a342c844142.png">

FirefoxMetzger replied:
Alright, then I will aim to write the docstrings from the classes/objects perspective.

FirefoxMetzger commented on pygfx/helpers/_axes.py:
```diff
@@ -33,6 +45,7 @@ def __init__(self, size=1.0, thickness=2):
             dtype="f4",
         )
 
+        # TODO: maybe change to colors that are friendly for colorblind users.
```
Yes, I would not address this here, because the PR will get big enough as-is.

almarklein replied:
#435

FirefoxMetzger commented on pygfx/geometries/_text.py:
```diff
@@ -201,14 +237,21 @@ def __init__(
+        -------
+        screen_space : bool
+            If False, text size uses the unit of the local frame (e.g. cm).
+            Otherwise it is uses the logical screen's units (e.g. px).
```
I will add this one manually because the text block will have to be reformatted.

FirefoxMetzger commented:
🥵 okay. A little bigger than expected, but we are getting there. I went over the entire library and updated all the public docstrings to numpydoc format. I've also added the ones that were missing (which were a surprising few) and made sure that parameters (for functions or constructors) are documented.

I didn't update the docstrings for `pygfx.camera`, `pygfx.controllers`, and `pygfx.linalg`, because the former two are due for a larger refactor (iirc, @almarklein?) and the later is scheduled for deprecation.

What is missing now is to (1) find a way to get autosummary to generate a link of the form `pygfx.WorldObject` that actually links to `pygfx.objects.WorldObject`, (2) to create a long (ctrl+F -able) list of items in the public API, and (3) write a class template that hides attributes and methods from parent classes to declutter the docs a bit.

Another point (that I wasn't quite clear on) is if we should also document internal classes in the generated documentation. These would live as part of the respective sub-package docs, but won't be listed in the public API. If so, we may wish to do this in a follow-up PR though because this one is getting a bit large.

almarklein commented:
<img width="159" alt="image" src="https://user-images.githubusercontent.com/3015475/213405544-2eae6d71-e5a9-4bc8-9bcd-d20c8c98a1fe.png">

🤯 

> because the former two are due for a larger refactor (iirc, @almarklein?) 

Yes, don't waste your time on these :)

> Another point (that I wasn't quite clear on) is if we should also document internal classes in the generated documentation.

I think it's enough if code for internal use is documented simply by having docstring that can be read in the source code. At least for the time being.

Korijn commented on docs/reference.rst:
```diff
@@ -50,119 +72,9 @@ Camera objects
     :member-order: bysource
```
I guess you will move the Camera objects to pygfx.cameras subpage?

FirefoxMetzger replied:
@Korijn  Yes that is the plan. I didn't touch them yet because we said they are due for a refactor, but I will add the plumbing  to keep the current classes documented.

Korijn replied:
I think that was about Controllers, not Cameras, actually. :)

almarklein replied:
Both, to some extend, probably ...

Korijn commented on pygfx/pkg_resources/__init__.py:
```diff
@@ -0,0 +1 @@
+# @almarklein: why do we need this init file here?
```
It relates to the packaging setup. This folder needs to be recognized as a python module. We load the binaries from this folder via package_resources, and it resolves paths by module names.

FirefoxMetzger replied:
Ah right. There is some logic in `pygfx.utils._dirs` that relies on the folder being a module to access resources. Thanks for clarifying

Korijn commented:
Nice work!!

> What is missing now is to (1) find a way to get autosummary to generate a link of the form `pygfx.WorldObject` that actually links to `pygfx.objects.WorldObject`, (2) to create a long (ctrl+F -able) list of items in the public API, and (3) write a class template that hides attributes and methods from parent classes to declutter the docs a bit.

In all three cases, if it ends up being costly (in terms of time and complexity), I am fine with the current output. Regarding (2):

There is a search bar on the documentation site and it works well. So it's not impossible to search for things. I don't think this is super important.

> Another point (that I wasn't quite clear on) is if we should also document internal classes in the generated documentation. These would live as part of the respective sub-package docs, but won't be listed in the public API. If so, we may wish to do this in a follow-up PR though because this one is getting a bit large.

I agree with Almar's reply to this.

FirefoxMetzger commented:
Okay, I resolved (2) and (3), and found a solution for (1) (see https://github.com/pygfx/pygfx/pull/430#issuecomment-1387739350  for details on what the individual points mean). Once we have resolved the current discussions this PR is RTR.

Korijn commented:
I'll be out on holiday between 26 jan and 7 feb, so I'll have limited access over my phone to approve PRs.

almarklein commented on pygfx/resources/_texture.py:
```diff
@@ -275,27 +285,45 @@ def format_from_memoryview(mem, size):
+        The dimensionality of the array (1, 2 or 3). If None, use the texture's
+        dimension. Or e.g. get a "2d" slice view from a 3d texture, or e.g.
+        "cube" or "2d-array".
+    aspect : str
+        Omit or pass None to use the default.
+        @almarklein: What does this do?
```
It's a (rather low level) wgpu thing. I don't think we use it somewhere, but for now it's not in the way either...
```suggestion
    aspect : Enum
        The `wgpu.TextureAspect` for this view. Omit or pass None to use the default.
```

almarklein commented on pygfx/utils/__init__.py:
```diff
@@ -1,3 +1,21 @@
+    show.Display
+    viewport.Viewport
+    text.font_manager
+    cm
```
In the docs there is also a `functions` section?  https://pygfx--430.org.readthedocs.build/en/430/_autosummary/pygfx.utils.html

almarklein replied:
Also the docs on e.g. `cm` speak of an alias: https://pygfx--430.org.readthedocs.build/en/430/_autosummary/utils/pygfx.utils.cm.html#pygfx.utils.cm

FirefoxMetzger replied:
> In the docs there is also a functions section?

True, looks like the default template adds a block for functions defined on a module. I'll have a look and remove them.

> Also the docs on e.g. cm speak of an alias:

Good catch. this should be a neatly rendered page of a sub-module.

FirefoxMetzger replied:
fixed.

almarklein commented on pygfx/utils/color.py:
```diff
@@ -33,19 +33,19 @@ def _float_from_css_value(v, i):
 
-    The above variations can also be supplied as a single tuple/list,
-    or anything that :
+    The above variations can also be supplied as a single tuple/list, or
+    anything that:
```
The bullet lists in this docstring don't become bullets in the rendered docs. Must they be indented?

FirefoxMetzger replied:
Yep, they should be :)

almarklein commented on pygfx/materials/_line.py:
```diff
@@ -101,26 +124,41 @@ def map(self, map):
     inconsistent and looks *very* thin on HiDPI monitors.
     """
 
+    # @almarklein: The docs advise to avoid this. Depricate before 1.0?
+
```
It may still be useful because the thin lines are more performant, or for e.g. debugging, so I'm reluctant to remove them.

FirefoxMetzger replied:
No problem. In that case, I'd say we just add that info to the docs so it is clear why it exists.

almarklein commented on pygfx/materials/_line.py:
```diff
@@ -101,26 +124,41 @@ def map(self, map):
+
+    A material that renders line segments that look like little vectors."""
+
+    # @almarklein: This is a thin (1:1) wrapper around LineMaterial. Out of
+    # curiosity, how does it do more than the base class?
```
Because the shader associated with this material does something different.

FirefoxMetzger replied:
Ah! so all the classes here that have an empty body have a shader in `pygfx.renderers.wgpu` that does something special when it encounters this class? I think this is something we should document in the narrative docs.

almarklein replied:
This is how it works for all materials, yes - the shader is selected based on the class of the material, and a shader may also make choices based on the subclass and/or their properties.

almarklein commented on pygfx/objects/_events.py:
```diff
@@ -118,13 +137,63 @@ def _update_current_target(self, target):
+    y : int
+        Thy y position of the cursor or touch in screen space (px).
+    button : int
+        @almarklein: where can I find the description of this attribute?
```
This is the spec that we follow for values like these: https://jupyter-rfb.readthedocs.io/en/latest/events.html

FirefoxMetzger replied:
I'll add the link to the docstring.

almarklein commented on pygfx/objects/_events.py:
```diff
@@ -176,13 +245,49 @@ def copy(self, **kwargs):
+    height : int
+        The new height of the application window in screen space (px).
+    pixel_ratio : float
+        The new pixel ratio. (@almarklein: is this aspect ratio or dpi or something else?)
```
```suggestion
        The new pixel ratio (ratio between logical pixels and physical pixels).
```

almarklein reviewed (commented):
I left a ton of comments. I think I got all the `@almarklein` refs 😄

FirefoxMetzger commented on pygfx/controllers/__init__.py:
```diff
@@ -1,8 +1,17 @@
 Controllers define how cameras can be interacted with. They are
 independent of any GUI toolkits.
+
+.. autosummary::
+
+    Controller
+    PanZoomController
+    OrbitController
+    OrbitOrthoController
```
good catch!

FirefoxMetzger commented on pygfx/geometries/_base.py:
```diff
@@ -6,36 +6,19 @@
 
 class Geometry(Trackable):
-    """A Geomerty object is a container for geometry data of a WorldObject.
+    """Base class for Geometry Subtypes.
```
```suggestion
    """Generic class for Geometry Subtypes.
```

@almarklein What about this instead?

almarklein replied:
Yes, that sounds good! 👍

FirefoxMetzger commented on pygfx/objects/_more.py:
```diff
@@ -4,21 +4,73 @@
+
+    Parameters
+    ----------
+    geometry : Geometry
+        The data defining the shape of the object.
+    material : Material
+        The data defining the appearence of the object.
+    visible : bool
+        Whether the object is visible.
+    render_order : int
+        The render order (when applicable for the renderer's blend mode).
+    render_mask : str
+        Determines the render passes that the object is rendered in. It's
+        recommended to let the renderer decide, using "auto".
+    position : Vector
+        The position of the object in the world. Default (0, 0, 0).
```
Mmh. In this case, we might also want to have a stub constructor that hides the arguments we remove here. Otherwise, they will still show up in an IDE and be suggested for auto-completion which may be confusing to newcomers.

FirefoxMetzger commented on pygfx/objects/_more.py:
```diff
@@ -28,51 +80,125 @@ def __init__(self, geometry=None, material=None, **kwargs):
 
 class Points(WorldObject):
-    """An object consisting of points represented by vertices (3D positions).
+    """A point cloud.
+
+    An object consisting of points represented by vertices (3D positions).
```
I think this is true for any object that is rendered. Ultimately, it is just a set of vertices with some extra data defined on them x)

FirefoxMetzger replied:
```suggestion
    """An object representing a disconnected cluster of individual points.

```

Maybe like this?

FirefoxMetzger replied:
> Is there a guideline on how short/long the "title" should be?

Yes there is. They should be a one-liner and not use the function name ([ref](https://numpydoc.readthedocs.io/en/latest/format.html#short-summary)). The main thing is to have something snappy that looks good when looking through the table in the docs. Otherwise we can do whatever we think is meaningful.

> I think some are so short they barely add anything over the class name.

Yes. sometimes its hard to say something useful beyond a descriptive class name. (guess it means you chose good variable names :D). It's just odd to not have a short-description because then there is a table entry that is randomly empty, which looks like we forgot something.

Korijn replied:
> Maybe like this?

I'd say the original was more understandable for a broader audience.

FirefoxMetzger replied:
@Korijn The original as in the version that is currently in the PR, or the version that Almar suggested?

Korijn replied:
> the version that is currently in the PR

:)

almarklein replied:
👍

FirefoxMetzger commented on pygfx/resources/_buffer.py:
```diff
@@ -6,27 +6,37 @@
 
 
 class Resource(Trackable):
+    """Resource base class."""
+
```
Yes, we can exclude the object from backreferencing, meaning that no objects will be shown. There might also be a way to only show a select number. I can investigate.

FirefoxMetzger replied:
Okay, I disabled indirect backreferences for `pygfx.Resource`. Now, only examples that link to it during the narrative comments of the example will be included.

FirefoxMetzger commented on pygfx/resources/_buffer.py:
```diff
@@ -6,27 +6,37 @@
+    storage buffer. You can provide (and update data for it), or use it as a
+    placeholder for a buffer with no representation on the CPU.
+
+    @almarklein: does this mean we differentiate between paged, pinned, and
+    device (GPU) memory?
+
```
Okay, so we don't differentiate explicitly in pygfx? I'm asking because this is a very big topic for deep learning and working with CUDA and I would have expected something similar for rendering.

If pygfx doesn't deal with this concept internally (or expose it in the Buffer class) then we can just leave it for now.

almarklein replied:
There is no way to specify these things in wgpu. We can only select whether it's a uniform, vertex, read-only-storage, regular storage. And we can specify where/how the buffer is used.

FirefoxMetzger commented on pygfx/geometries/__init__.py:
```diff
@@ -1,3 +1,70 @@
+    :toctree: geometry/
+    :template: ../_templates/clean_class.rst
+
+    trimesh_geometry
+    octahedron_geometry
+    icosahedron_geometry
+    dodecahedron_geometry
+    tetrahedron_geometry
+    torus_knot_geometry
+    klein_bottle_geometry
```
Nice catch. I thought that autosummary would be smart enough to use the function template for functions and class template for classes. However, it looks like you always use the custom template if you specify a custom template.

I've updated it to treat functions correctly

FirefoxMetzger commented on pygfx/materials/__init__.py:
```diff
@@ -1,3 +1,52 @@
+    pillow_image
+    trimesh_material
+
+    # @almarklein the descriptions for the materials are sparse, but I lack
+    # context to come up with something more precise. Could you make a suggestion?
```
applied.

FirefoxMetzger commented on pygfx/materials/_mesh.py:
```diff
@@ -6,10 +6,53 @@
+
     """
 
+    # @almarklein: Could we be more precise on what "physical space" means for
+    # the map's color format?
```
I'll add a link in the docs.

FirefoxMetzger commented on pygfx/objects/_more.py:
```diff
@@ -4,21 +4,73 @@
-    """An object representing a scene background.
+    """The scene's background.
+
+    @almarklein: how come this is an independent object and not part of Scene?
```
okay. If this is already tracked then let's not make another issue for it.

FirefoxMetzger commented on pygfx/resources/_texture.py:
```diff
@@ -4,27 +4,37 @@
+            derived from dim and the shape of the data. By creating a 2D array
+            with ``depth > 1``, a view can be created with format 'd2_array' or
+            'cube'.
+            @almarklein: I don't understand this behavior. Can you elaborate so
+            I can come up with a better docstring?
```
@almarklein This is the final discussion point. I'm not exactly sure what the second sentence (about the 2D and cube array means).

FirefoxMetzger replied:
We are down to one unresolved conversation. Once this one is cleared the PR is ready :)

FirefoxMetzger commented on pygfx/objects/_more.py:
```diff
@@ -4,21 +4,73 @@
+
+    Parameters
+    ----------
+    geometry : Geometry
+        The geometry to use when rendering the background.
```
Seeing this in the code now, it looks like we allow `geometry` for backwards compatibility:

https://github.com/pygfx/pygfx/blob/774a983c0133f458e502b1eef34398f1e8174b93/pygfx/objects/_more.py#L24-L27

Is this safe to remove?

almarklein replied:
I'm leaning towards forcing `Background(None, material)`, because all world objects specify geometry and material in that order. Let's not spend to much time on this though, since the background is a bit up in the air anyway.

FirefoxMetzger replied:
okay. in this case I will rephrase the docstring instead of removing it.

almarklein commented on docs/conf.py:
```diff
@@ -85,6 +56,7 @@ def _check_objects_are_documented():
     "examples_dirs": "../examples",
     "gallery_dirs": "_gallery",
     "backreferences_dir": "_gallery/backreferences",
+    "exclude_implicit_doc": {r"Resource"},
```
Exclude some more, e.g. base classes.
```suggestion
    "exclude_implicit_doc": {"WgpuRenderer", "Resource", "WorldObject", "Geometry", "Material", "Controller", "Camera", "show", "Display", "Group", "Scene", "Light"},
```

almarklein commented on pygfx/resources/_texture.py:
```diff
@@ -4,27 +4,37 @@
+            The data is copied if it's not float32 or not contiguous.
+        dim : int
+            The dimensionality of the array (1, 2 or 3).
+        size : tuple, [3]
+            The extent ``(width, height, depth)`` of the array. If None, it is
+            derived from dim and the shape of the data. By creating a 2D array
+            with ``depth > 1``, a view can be created with format 'd2_array' or
+            'cube'.
+            @almarklein: I don't understand this behavior. Can you elaborate so
+            I can come up with a better docstring?
```
```suggestion
        size : tuple, [3]
            The extent ``(width, height, depth)`` of the array. If None, it is
            derived from `dim` and the shape of the data. The texture can also
            represent a stack of images by setting `dim=2` and `depth > 1`.
            Any derived texture views must then have a `view_dim` of  e.g. 'd2_array' or 'cube'.
```

FirefoxMetzger replied:
👍

almarklein reviewed (commented):
Almost done!

FirefoxMetzger commented:
> Almost done!

@almarklein Done? 😅 

(This PR is getting too long for GH. My browser is becoming slow when I search for open discussions :D)

almarklein reviewed (approved):
Done!

Korijn commented:
Amazing work!
